beacon-chain/core/blocks/block_operations.go

```diff
@@ -84,3 +89,183 @@ func verifyProposerSlashing(
 	}
 	return nil
 }
+
+// ProcessCasperSlashings is one of the operations performed
+// on each processed beacon block to penalize validators based on
+// Casper FFG slashing conditions if any slashable events occurred.
+//
+// Official spec definition for casper slashings:
+// Verify that len(block.body.casper_slashings) <= MAX_CASPER_SLASHINGS.
+//   For each casper_slashing in block.body.casper_slashings:
+//
+//   Verify that verify_casper_votes(state, casper_slashing.votes_1).
+//   Verify that verify_casper_votes(state, casper_slashing.votes_2).
+//   Verify that casper_slashing.votes_1.data != casper_slashing.votes_2.data.
+//   Let indices(vote) = vote.aggregate_signature_poc_0_indices +
+//     vote.aggregate_signature_poc_1_indices.
+//   Let intersection = [x for x in indices(casper_slashing.votes_1)
+//     if x in indices(casper_slashing.votes_2)].
+//   Verify that len(intersection) >= 1.
+//   Verify that casper_slashing.votes_1.data.justified_slot + 1 <
+//     casper_slashing.votes_2.data.justified_slot + 1 ==
+//     casper_slashing.votes_2.data.slot < casper_slashing.votes_1.data.slot
+//     or casper_slashing.votes_1.data.slot == casper_slashing.votes_2.data.slot.
+//   For each validator index i in intersection,
+//     if state.validator_registry[i].status does not equal
+//     EXITED_WITH_PENALTY, then run
+//     update_validator_status(state, i, new_status=EXITED_WITH_PENALTY)
+func ProcessCasperSlashings(
+	validatorRegistry []*pb.ValidatorRecord,
+	casperSlashings []*pb.CasperSlashing,
+	currentSlot uint64,
+) ([]*pb.ValidatorRecord, error) {
+	if uint64(len(casperSlashings)) > params.BeaconConfig().MaxCasperSlashings {
+		return nil, fmt.Errorf(
+			"number of casper slashings (%d) exceeds allowed threshold of %d",
+			len(casperSlashings),
+			params.BeaconConfig().MaxCasperSlashings,
+		)
+	}
+	for idx, slashing := range casperSlashings {
+		// verifyCasperSlashing performs validity checks on the slashing
+		// struct and returns a list of validator indices to be slashed from
+		// the BeaconState's validator registry.
+		validatorIndices, err := verifyCasperSlashing(slashing)
+		if err != nil {
+			return nil, fmt.Errorf("could not verify casper slashing #%d: %v", idx, err)
+		}
+		for _, validatorIndex := range validatorIndices {
+			proposer := validatorRegistry[validatorIndex]
```

Comment:
Nitpick, I don't think `propoer` is a good variable name here. Consider `validator` or `penalizedValidator`